### PR TITLE
feat(dflash): add robust Laguna target and draft support

### PR DESCRIPTION
**Pull request by FabioMalpezzi**
## Acknowledgement

First of all, many thanks to **Jun Kim**, the maintainer of this repository, for the excellent work on oMLX. The project has been extremely useful to me, and its architecture and integration points are thoughtful and well designed. I hope this contribution is useful to the project in return.

## Summary

This PR extends the existing DFlash engine to support Poolside Laguna targets and their official DFlash draft checkpoints.

- register a strict, idempotent Laguna `TargetOps` backend through dflash-mlx's existing extension points;
- add the gated Laguna draft specialization used by Poolside checkpoints;
- support fused-QKV checkpoint loading, partial RoPE, per-head/per-element attention gates, captured-layer RMS normalization, mixed full/sliding target caches, and rejection rollback;
- validate target/draft family, depth, hidden size, and capture-layer compatibility at load time;
- explicitly bind every DFlash draft to its target, preserving the target embedding scale for Gemma4 while remaining a no-op for Qwen and Laguna;
- apply the existing MoE gate+up fusion path to compatible DFlash targets;
- expose Laguna as a supported DFlash family in the admin UI and documentation;
- add Laguna adapter, rollback, checkpoint-layout, configuration, and prefix-cache round-trip coverage.

## Design priority

While implementing this PR, I deliberately prioritized **robust inference and exact target-model behavior over absolute peak performance**.

For that reason, DDTree verification, target KV quantization, and the specialized verify-linear path remain disabled for Laguna until each has dedicated numerical-parity coverage. The implementation uses ordinary adaptive DFlash verification and fails early on incompatible target/draft combinations instead of attempting a permissive fallback.

### oMLX tests

```text
178 passed in 1.63s
```

The run covered:

- `tests/test_dflash_engine.py`
- `tests/test_dflash_laguna.py`
- `tests/test_dflash_lifecycle.py`
- `tests/test_qwen35_moe_gate_up.py`
- `tests/test_dflash_prefill_memory_guard.py`
- `tests/test_model_settings.py`

### dflash-mlx adapter tests

```text
67 passed, 5 skipped in 2.01s
```

The five skipped tests are upstream opt-in real-model tests that expect their default models in the Hugging Face snapshot cache. Real local-model parity for the selected Laguna, Gemma4, and Qwen3.6 pairs was exercised directly by the benchmarks below; every run produced the exact same greedy token sequence as native generation.

### Machine

- MacBook Pro (`Mac16,6`)
- Apple M4 Max
- 16-core CPU: 12 performance + 4 efficiency cores
- 40-core integrated GPU, Metal 4
- 128 GB unified memory
- macOS 26.5.1 (`25F80`)
- Python 3.12.11
- MLX 0.32.0
- mlx-lm 0.31.3
- dflash-mlx 0.1.10
- oMLX 0.5.3, branch head `39c8a016`

### Method

- greedy decoding (`temperature=0`);
- same Fibonacci coding prompt for native and DFlash paths;
- 64 generated tokens;
- one warm-up before measured runs;
- median of 5 measured runs for Laguna and 3 measured runs for Gemma4/Qwen3.6;
- block size 16, `verify_mode=dflash`, CopySpec off;
- draft quantization `w4a16:gs64`;
- prefix cache disabled to isolate single-request generation;
- exact generated-token parity required.

| Target / draft | Native wall | DFlash wall | Native tok/s | DFlash tok/s | Speedup | Acceptance | Exact parity | Peak memory |
|---|---:|---:|---:|---:|---:|---:|:---:|---:|
| `poolside/Laguna-S-2.1-NVFP4-mlx` / `Laguna-S-2.1-DFlash-NVFP4` | 1.968 s | 0.970 s | 32.5 | 66.0 | **2.03x** | 90.6% | yes | 72.69 GB |
| `mlx-community/gemma-4-26b-a4b-it-4bit` / `z-lab/gemma-4-26B-A4B-it-DFlash` | 0.722 s | 0.378 s | 88.6 | 169.2 | **1.91x** | 90.6% | yes | 14.58 GB |
| `mlx-community/Qwen3.6-35B-A3B-4bit` / `z-lab/Qwen3.6-35B-A3B-DFlash` | 0.650 s | 0.262 s | 98.5 | 244.2 | **2.48x** | 92.2% | yes | 19.94 GB |

The cross-family runs also verified the shared changes:

- Gemma4 draft `embed_scale` was bound from `1.0` to the target value `53.066`, with exact output parity;
- Qwen retained `embed_scale=1.0`, and gate+up fusion applied to all 40 MoE layers;
- the Laguna adapter fused 47 MoE layers in the latest rerun;
- Gemma4 remained outside the MoE fusion family filter, as intended.

> Note: the current Qwen3.6 drafter checkpoint was updated to a Transformers 5.7-style config where `block_size` and `rope_theta` are nested under `dflash_config` and `rope_parameters`. dflash-mlx 0.1.10 still expects root-level aliases, so the local benchmark added equivalent root aliases (`16` and `10000000`) without changing weights or semantics. This is an upstream checkpoint/runtime metadata compatibility issue and is not part of this PR.

## Intentionally unsupported Laguna fast paths

- DDTree / tree verification;
- target KV-cache quantization;
- specialized verify-linear replacement.

These can be enabled incrementally once model-specific parity tests demonstrate that rollback, cache state, and logits remain exact.

### CODE REVIEW

**Comment by jundot** [comment]
Thanks for putting this together. I added a prefix-cache fix so sparse `TargetHiddenChunks` are normalized chunk-by-chunk while preserving their spans, including the projected-context cache advance path. I verified 48-token greedy parity with a real Laguna S 2.1 FP8 target, and a 922-token trimmed snapshot hit now matches the cold path. LGTM after this fix; merging.

**Comment by FabioMalpezzi** [comment]
Thank you very much for the quick response and for adding this fix. Having proper Laguna support directly in the project’s main branch will be extremely useful in my work.

I recently retired from my role as a CIO/CEO and now work as an independent AI consultant. This gives me more time and flexibility to contribute to interesting open-source projects.

If you ever need any help with oMLX support, testing, documentation, or any other aspect of the project, I would be more than happy to contribute.
